ray-operator/controllers/ray/batchscheduler/koordinator/koordinator_scheduler_test.go

```diff
@@ -0,0 +1,103 @@
+package koordinator
+
+import (
+	"context"
+	"testing"
+
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+	"github.com/ray-project/kuberay/ray-operator/controllers/ray/utils"
+	"github.com/stretchr/testify/assert"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func createRayClusterWithLabels(namespace, name string, labels map[string]string) *rayv1.RayCluster {
+	rayCluster := &rayv1.RayCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels:    labels,
+		},
+	}
+
+	return rayCluster
+}
+
+func setHeadPodNamespace(app *rayv1.RayCluster, namespace string) {
+	app.Spec.HeadGroupSpec.Template.Namespace = namespace
+}
+
+func addWorkerPodSpec(app *rayv1.RayCluster, workerGroupName string,
+	namespace string, replicas, minReplicas int32) {
+	workerGroupSpec := rayv1.WorkerGroupSpec{
+		GroupName:   workerGroupName,
+		Replicas:    &replicas,
+		MinReplicas: &minReplicas,
+	}
+	workerGroupSpec.Template.Namespace = namespace
+
+	app.Spec.WorkerGroupSpecs = append(app.Spec.WorkerGroupSpecs, workerGroupSpec)
+}
+
+func createPod(namespace, name string) *v1.Pod {
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   namespace,
+			Labels:      make(map[string]string),
+			Annotations: make(map[string]string),
+		},
+	}
+}
+
+func TestAddMetadataToPod(t *testing.T) {
+	ks := &KoodinatorScheduler{}
+	ctx := context.Background()
+	// test the case when gang-scheduling is enabled
+	rayClusterWithGangScheduling := createRayClusterWithLabels(
+		"ray-namespace",
+		"koord",
+		map[string]string{
+			utils.RayClusterGangSchedulingEnabled: "true",
+		},
+	)
+
+	setHeadPodNamespace(rayClusterWithGangScheduling, "ns0")
+	addWorkerPodSpec(rayClusterWithGangScheduling, "workergroup1", "ns1", 4, 2)
+	addWorkerPodSpec(rayClusterWithGangScheduling, "workergroup2", "ns2", 5, 3)
```

Comment:
Does Koordinator support having RayCluster pods spread across different namespaces? I don't think that's really supported with Kuberay

Reply:
> Does Koordinator support having RayCluster pods spread across different namespaces? I don't think that's really supported with Kuberay

As mentioned in the documentation at https://koordinator.sh/docs/designs/gang-scheduling/#annotation-way, the Koordinator scheduler supports highly flexible GangGroup scheduling requirements, including the ability to support different pods in a GangGroup belonging to different namespaces.

Of course, if you believe such special scheduling requirements are unlikely to occur in KubeRay, I fully agree as well.